src/parcsr_mv/par_csr_filter.c

```diff
@@ -157,27 +162,48 @@ hypre_ParCSRMatrixBlkFilterHost( hypre_ParCSRMatrix  *A,
       B_offd_i[i + 1] = B_offd_i[i];
       for (j = A_offd_i[i]; j < A_offd_i[i + 1]; j++)
       {
-         if (c == (HYPRE_Int) (col_map_offd_A[A_offd_j[j]] % big_block_size))
+         big_col = col_map_offd_A[A_offd_j[j]];
+         if (c == (HYPRE_Int) (big_col % big_block_size))
          {
-            B_offd_j[B_offd_i[i + 1]] = A_offd_j[j];
-            B_offd_a[B_offd_i[i + 1]] = A_offd_a[j];
+            B_offd_bj[B_offd_i[i + 1]] = big_col;
+            B_offd_a[B_offd_i[i + 1]]  = A_offd_a[j];
             B_offd_i[i + 1]++;
-            marker[A_offd_j[j]] = 1;
          }
       }
    }
 
-   /* Update col_map array */
-   num_cols_offd_B = 0;
-   for (i = 0; i < num_cols_offd_A; i++)
+   /* Allocate work array */
+   work = hypre_TAlloc(HYPRE_BigInt, B_offd_nnz, HYPRE_MEMORY_HOST);
+   hypre_TMemcpy(work, B_offd_bj, HYPRE_BigInt, B_offd_nnz,
+                 HYPRE_MEMORY_HOST, HYPRE_MEMORY_HOST);
+
+   /* Remove duplicate columns */
+   hypre_BigQsort0(work, 0, B_offd_nnz - 1);
+   num_cols_offd_B = (B_offd_nnz == 0) ? 0 : 1;
+   for (i = 0; i < B_offd_nnz - 1; i++)
    {
-      if (marker[i])
+      if (work[i + 1] > work[i])
       {
-         col_map_offd_B[num_cols_offd_B++] = col_map_offd_A[i];
+         work[num_cols_offd_B++] = work[i + 1];
       }
    }
+
+   /* Build B's col_map array */
+   for (i = 0; i < num_cols_offd_B; i++)
+   {
+      col_map_offd_B[i] = work[i];
```

Comment:
Why a for loop instead of just a memcpy?

Reply:
That would work as well! I picked the for loop option because `num_cols_offd_B` is generally small and the compiler can do efficient loop unrolling strategies that might be faster than memcpy. But the difference is very minimal to be honest...

Reply:
I didn't realize a for loop might actually be faster than memcpy! Anyway, doesn't matter to me either way, I guess, you can do either here.